### PR TITLE
NUI-1650, NUI-1214 implement smarter retry logic for 429s in asset-compute-client

### PR DESCRIPTION
**Pull request by jdelbick**
## Description

We have throttling limits per client (150 req per min) and per org (300 req per min) in our service that we often hit in integration tests or load tests. This new feature offers an alternative client `AssetComputeClientRetry` that is a lightweight wrapper around `AssetComputeClient` with retry logic for 429s.

`API gateway`, the home of our asset compute service endpoints, sends back a `retry-after` header when a request is rejected with 429s: https://wiki.corp.adobe.com/pages/viewpage.action?pageId=1398492571

The new client uses this header (+small buffer time) to determine how long to wait between retries on 429s. 

Fixes # https://jira.corp.adobe.com/browse/NUI-1650
related issue: https://jira.corp.adobe.com/browse/NUI-1214
(adapted code from here: https://git.corp.adobe.com/nui/integration-tests/blob/72dfbb4de0387486afe77f191fc2ec48a5bd9863/lib/asset-compute-retry.js)

### Testing
I have tested this running ITs concurrently locally (using npm linking). I tested with 3 IT tests concurrently. Results were much better than randomly choosing numbers between 30-60s to wait and tests completed faster. Last test finished in ~4.3mins
```
[0]     >>> No longer using integration 1
[0]     integration pool stopped
[0] Test set complete. Took 258.561s
```
There was a range of `retry-after` times (~1 - >20s)
I plan to test with more concurrently (up to 7) during non-working hours

### Design decisions
I chose to make this a wrapper (extension of the class) instead of an `opt-in` for a few reasons:
1. This was originally created outside of the library (in integration tests)
2. Implementing this inside `AssetComputeClient` (client.js) as an opt-in would have added a bunch of code around each request that is only used sometimes. (code readability)
3. It will be very obvious to the user of the library which version they will be using
4. You only need to change the code at the import level to use the retry version

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -15,11 +15,13 @@
 const { AssetCompute } = require("./lib/assetcompute");
 const { AssetComputeEventEmitter } = require("./lib/eventemitter");
 const { AssetComputeClient } = require("./lib/client");
+const { AssetComputeClientRetry } = require("./lib/client-retry");
```
Comment (jdelbick) [inline index.js:18]: let me know if you prefer a different name, something that clearly lets the user know this is related to retries on 429s: `AssetComputeClient429Retry`

File: lib/assetcompute.js [hunk 2]
```diff
@@ -117,7 +121,11 @@ class AssetCompute {
         });
         if (!response.ok) {
             const msg = await response.text();
-            throw Error(`Unable to invoke /unregister: ${response.status}: ${msg} (details: ${JSON.stringify(response)})`);
+            const error = new Error(`Unable to invoke /unregister: ${response.status}: ${msg} (details: ${JSON.stringify(response)})`);
+            if (response.status === 429) {
+                error.retryAfter = parseInt(response.headers.get('retry-after')); 
+            }
+            throw error;
```
Comment (jdelbick) [inline lib/assetcompute.js:128]: the response headers were not sent back to the client since `!response.ok` was sent back as an error, so I added the `retry-after` header to the error.

An improvement would be to create a custom error class for these 429 errors

File: lib/client-retry.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+ */
+
+"use strict";
+const { AssetComputeClient } = require("./client");
+const colors = require("colors/safe");
+const sleep = require("util").promisify(setTimeout);
+
+const MAX_RETRY_429 = 4;
+const MAX_RETRY_WAIT_TIME = 60000;
+const MIN_RETRY_WAIT_TIME = 30000;
```
Comment (jdelbick) [inline lib/client-retry.js:20]: future improvement: 
- make the backup waittime configurable (if it even would be used)
- remove the backup waittime (if its not ever used)

File: index.js (hunk 1 above)
Comment (tmathern) [reply index.js:18]: No, in case we ever need to extend. This name is fine!

File: lib/assetcompute.js [hunk 4]
```diff
@@ -117,7 +121,11 @@ class AssetCompute {
         });
         if (!response.ok) {
             const msg = await response.text();
-            throw Error(`Unable to invoke /unregister: ${response.status}: ${msg} (details: ${JSON.stringify(response)})`);
+            const error = new Error(`Unable to invoke /unregister: ${response.status}: ${msg} (details: ${JSON.stringify(response)})`);
+            if (response.status === 429) {
+                error.retryAfter = parseInt(response.headers.get('retry-after'));
```
Comment (tmathern) [inline lib/assetcompute.js:126]:
```suggestion
                error.retryAfter = parseInt(response.headers.get('retry-after'), 10); 
```

File: lib/assetcompute.js [hunk 5]
```diff
@@ -162,7 +170,11 @@ class AssetCompute {
         });
         if (!response.ok) {
             const msg = await response.text();
-            throw Error(`Unable to invoke /process: ${response.status} ${msg}`);
+            const error = new Error(`Unable to invoke /process: ${response.status} ${msg}`);
+            if (response.status === 429) {
+                error.retryAfter = parseInt(response.headers.get('retry-after'));
```
Comment (tmathern) [inline lib/assetcompute.js:175]:
```suggestion
                error.retryAfter = parseInt(response.headers.get('retry-after'), 10); 
```

File: lib/client-retry.js [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+ * governing permissions and limitations under the License.
+ */
+
+"use strict";
+const { AssetComputeClient } = require("./client");
+const colors = require("colors/safe");
+const sleep = require("util").promisify(setTimeout);
+
+const MAX_RETRY_429 = 4;
+const MAX_RETRY_WAIT_TIME = 60000;
```
Comment (tmathern) [inline lib/client-retry.js:19]:
```suggestion
const MAX_RETRY_WAIT_TIME_MS = 60000;
```

File: lib/client-retry.js (hunk 3 above)
Comment (tmathern) [reply lib/client-retry.js:20]:
```suggestion
const MIN_RETRY_WAIT_TIME_MS = 30000;
```

File: lib/client-retry.js (hunk 6 above)
Comment (tmathern) [reply lib/client-retry.js:19]: (unit)

File: lib/client-retry.js (hunk 3 above)
Comment (tmathern) [reply lib/client-retry.js:20]: (unit)

File: lib/client-retry.js [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    constructor(integration, options) {
+        super(integration, options);
+        this.max429RetryCount = (options && options.max429RetryCount) || MAX_RETRY_429;
+    }
+
+    retryWaitTime(e) {
+        // randomly choose between 30-60 seconds waittime
+        let waitTime = Math.floor(Math.random() * (MAX_RETRY_WAIT_TIME - MIN_RETRY_WAIT_TIME + 1)) + MIN_RETRY_WAIT_TIME;
+
+        // in seconds amount of time to wait before retrying the request
```
Comment (tmathern) [inline lib/client-retry.js:35]: seconds or milliseconds? I have a doubt now - seems large to be in seconds

File: lib/client-retry.js [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+                    console.log(colors.red(`Waiting for ${waitTime}ms and trying again`));
+                    await sleep(waitTime);
+                } else {
+                    non429Response = true;
+                    error = e;
+                }
+            }
+        }
+        if (!non429Response) {
+            throw new Error("Running into 429s. Will not continue testing. Try re-running in a few minutes.");
```
Comment (tmathern) [inline lib/client-retry.js:64]: In case the library is not used in tests:
```suggestion
            throw new Error("Running into 429s. Will not continue retrying. Try again in a few minutes.");
```

File: lib/client-retry.js [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+                    await sleep(waitTime);
+                } else {
+                    non429Response = true;
+                    error = e;
+                }
+            }
+        }
+
+        if (!non429Response) {
+            throw new Error("Running into 429s. Will not continue testing. Try re-running in a few minutes.");
```
Comment (tmathern) [inline lib/client-retry.js:98]: In case the library is not used in tests:
```suggestion
            throw new Error("Running into 429s. Will not continue retrying. Try again in a few minutes.");
```

File: lib/client-retry.js [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
```
Comment (tmathern) [inline lib/client-retry.js:131]: In case the library is not used in tests:
```suggestion
            throw new Error("Running into 429s. Will not continue retrying. Try again in a few minutes.");
```

**Review by tmathern — CHANGES_REQUESTED** [review]

File: lib/client-retry.js [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+        let non429Response = false;
+        let retryCount = 0;
+        while (!non429Response && retryCount <= this.max429RetryCount) {
+            retryCount++;
+            try {
+                response = await super.register();
+                non429Response = true;
+            } catch (e) {
+                if (e.message.includes("429")) {
+                    console.log(colors.red(`encountered a 429: ${e.message}.`));
```
Comment (pheenomenon) [inline lib/client-retry.js:53]: it may seem unnecessary, but `non429Response = false` here could help with readability.

File: lib/client-retry.js (hunk 8 above)
Comment (pheenomenon) [inline lib/client-retry.js]:
```suggestion
            throw new Error(`Running into 429s after ${retryCount} attempts. Will not continue testing. Try re-running in a few minutes.`);
```

File: lib/client-retry.js (hunk 7 above)
Comment (jdelbick) [reply lib/client-retry.js:35]: sorry about the unclearness here
- api gateway `retry-after` header comes back in seconds
- every other variable is in ms

Comment (jdelbick) [reply lib/client-retry.js:35]: (I only know this from testing, still on the hunt for the official api gateway docs)

File: lib/client-retry.js [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+        }
+        return response;
+    }
+
+    // retry on 429s
+    async unregister() {
+        let response, error;
+        let non429Response = false;
+        let retryCount = 0;
+        while (!non429Response && retryCount <= this.max429RetryCount) {
```
Comment (pheenomenon) [inline lib/client-retry.js:79]: could be extracted out to reduce redundancy.

File: test/assetcompute.test.js [hunk 13]
```diff
@@ -48,6 +48,32 @@ describe( 'assetcompute.js tests', () => {
         assert.strictEqual(response.requestId, requestId);
         assert.strictEqual(response.journal, journal);
     });
+    it('should fail with asset compute /register hits 429', async function() {
```
Comment (pheenomenon) [inline test/assetcompute.test.js:51]: we dont have retry for this correct? can we mention it in the name ?
```suggestion
    it('should fail without retry when asset compute /register hits 429', async function() {
```

File: test/client-retry.test.js [hunk 14]
```diff
@@ -0,0 +1,843 @@
...
+        });
+        mockRequire.reRequire('../lib/client');
+        mockRequire.reRequire('../lib/client-retry');
+    });
+
+    afterEach(() => {
+        mockRequire.stopAll();
+        nock.cleanAll();
+    });
+    it('should create asset compute client with retry on 429s', async function () {
```
Comment (pheenomenon) [inline test/client-retry.test.js:82]:
```suggestion
    it('should create asset compute client with config to retry on 429s', async function () {
```

**Review by pheenomenon — APPROVED** [review]

File: lib/client-retry.js (hunk 7 above)
Comment (jdelbick) [reply lib/client-retry.js]: I guess I should take into account that it could be a date (says the doc), I will add this

**Review by tmathern — APPROVED** [review]

File: lib/client-retry.js (hunk 7 above)
Comment (jdelbick) [reply lib/client-retry.js]: Alright done: https://github.com/adobe/asset-compute-client/pull/53/commits/80f1650a6cbcb4002685ee0b065092c9283e5c37
I created a new custom error class: `TooManyRequestsError`
It handles converting the stringified `retry-after` header (either as a date or number) into a number in seconds to wait until retrying

**Comment by jdelbick** [comment]
@tmathern @pheenomenon I made some changes (adding the `TooManyRequestsError` and some tests), so I thought I would re-request review

File: index.js [hunk 15]
```diff
@@ -15,11 +15,15 @@
...
+const { TooManyRequestsError } = require('./lib/error');
 
 module.exports = {
     AssetCompute,
     AssetComputeEventEmitter,
     AssetComputeClient,
-    getIntegrationConfiguration
+    AssetComputeClientRetry,
+    getIntegrationConfiguration,
+    TooManyRequestsError
```
Comment (tmathern) [inline index.js:28]: Eventually, should make its way to the asset-compute-commons error file/lib.

File: lib/assetcompute.js [hunk 16]
```diff
@@ -161,8 +172,13 @@ class AssetCompute {
             retryOptions: this.retryOptions
         });
         if (!response.ok) {
-            const msg = await response.text();
-            throw Error(`Unable to invoke /process: ${response.status} ${msg}`);
+            const responseText = await response.text();
```
Comment (tmathern) [inline lib/assetcompute.js:175]: You could make those few lines of code a function with the invoked endpoint as parameter to avoid the code repetitions.

File: lib/client-retry.js [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+"use strict";
+const { AssetComputeClient } = require("./client");
+const colors = require("colors/safe");
+const sleep = require("util").promisify(setTimeout);
+
+const MAX_RETRY_429 = 4;
```
Comment (tmathern) [inline lib/client-retry.js:18]: Not sure, maybe a little bit more clearer... I'll let you judge.
```suggestion
const MAX_RETRY_ON_HTTP_429 = 4;
```

File: lib/client-retry.js [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+const MAX_RETRY_WAIT_TIME_MS = 60000; // maximum time in ms to wait before retrying if no `retry-after` header is sent
+const MIN_RETRY_WAIT_TIME_MS = 30000; // minimum time in ms to wait before retrying if no `retry-after` header is sent
+const TOO_MANY_REQUEST_ERROR = 'TooManyRequestsError';
+const TOO_MANY_REQUEST_ERROR_CODE = 429;
+/**
+ * Small wrapper around AssetComputeClient adding retry logic on 429s based on `retry-after` header
+ */
+class AssetComputeClientRetry extends AssetComputeClient {
```
Comment (tmathern) [inline lib/client-retry.js:26]:
```suggestion
class AssetComputeClientWithRetry extends AssetComputeClient {
```
Maybe?

File: lib/client-retry.js (hunk 12 above)
Comment (tmathern) [reply lib/client-retry.js:82]: +1

File: lib/client-retry.js [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+            throw new Error(`Running into 429s after ${retryCount} attempts. Will not continue retrying. Try again in a few minutes.`);
+
+        }
+        // pass through error
+        if (error) {
+            throw error;
+        }
+        return response;
+    }
+    async process(source, renditions, userData) {
```
Comment (tmathern) [inline lib/client-retry.js:112]:
```suggestion

    async process(source, renditions, userData) {
```
(it reads nicer)

**Review by tmathern — CHANGES_REQUESTED** [review]
The logic looks OK, but there is redundant code that should be refactored so it's easier to maintain (and understand for people who never worked on that code).

File: index.js (hunk 15 above)
Comment (jdelbick) [reply index.js:28]: I was thinking that, but it is very specific to the asset compute client (and the lib doesn't have a dep on commons yet)

File: lib/assetcompute.js (hunk 16 above)
Comment (jdelbick) [reply lib/assetcompute.js:167]: Done

File: lib/client-retry.js (hunk 12 above)
Comment (jdelbick) [reply lib/client-retry.js:82]: since its a `while` loop, each part needs to be re-checked dynamically. If I store it as a variable like the following, it doesn't change between rounds and ends up running forever:
```
shouldRetry = !non429Response && retryCount <= this.max429RetryCount;
while (shouldRetry) {
```
the other option I could do is put it in a function, but not sure if that makes the code cleaner:
```
while (shouldRetry(non429Response, retryCount)) {
```
wdyt?

File: test/assetcompute.test.js [hunk 20]
```diff
@@ -48,9 +48,166 @@ describe( 'assetcompute.js tests', () => {
         assert.strictEqual(response.requestId, requestId);
         assert.strictEqual(response.journal, journal);
     });
+    it('should calling asset compute /register with 401', async function() {
```
Comment (pheenomenon) [inline test/assetcompute.test.js:51]:
```suggestion
    it('should fail calling asset compute /register with 401', async function() {
```

File: lib/assetcompute.js [hunk 21]
```diff
@@ -116,8 +122,13 @@ class AssetCompute {
             }
         });
         if (!response.ok) {
-            const msg = await response.text();
-            throw Error(`Unable to invoke /unregister: ${response.status}: ${msg} (details: ${JSON.stringify(response)})`);
+            const responseText = await response.text();
+            const message = `Unable to invoke /unregister: ${response.status}: ${responseText} (details: ${JSON.stringify(response)})`;
```
Comment (pheenomenon) [inline lib/assetcompute.js:122]: Any reason why only /unregister has `(details: ${JSON.stringify(response)})` in message and not /register ?
`const message = `Unable to invoke /register: ${response.status} ${responseText}`;`

Comment (jdelbick) [reply lib/assetcompute.js:122]: I am honestly not sure... stringifying the entire `/process` response may be too long

File: test/assetcompute.test.js [hunk 22]
```diff
@@ -102,8 +259,35 @@ describe( 'assetcompute.js tests', () => {
...
+                'retry-after': 3
+            });
+
+        const assetCompute = new AssetCompute(options);
+        try {
+            await assetCompute.unregister();
+        } catch (error) {
+            assert.strictEqual(error.code, 429);
+            assert.strictEqual(error.name, 'TooManyRequestsError');
+            assert.strictEqual(error.message, 'Unable to invoke /unregister: 429: {"ok":false,"error_code":"429050","message":"Too many requests"} (details: {"size":0,"timeout":30000})');
```
Comment (pheenomenon) [inline test/assetcompute.test.js:285]: curious, where is `30000` coming from ?

File: lib/client-retry.js (hunk 12 above)
Comment (jdelbick) [reply lib/client-retry.js:82]: I am also trying to have a shared `retry` function in `client-retry.js` where I pass the `super.register()` or `super.process` function to the shared function to execute, but I am struggling to get it to work, any suggestions?
```
async retry(endpoint) {
   // ... retry stuff
   // call endpoint:
    response = await endpoint();
}
async register() {
    return this.retry(super.register);
}
```

File: test/assetcompute.test.js (hunk 22 above)
Comment (jdelbick) [reply test/assetcompute.test.js:285]: this is taken from real-world example of trying this out in IT tests and seeing this message. I am not totally sure as the api-gateway docs do not provide much details: https://wiki.corp.adobe.com/display/API/Error+Codes+Documentation+-+API+Gateway

File: lib/error.js [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+        // `retry-after` header can be a stringified number in seconds or a stringified date
+        if (typeof(retryAfterHeader) === 'string') {
+            if (isStringifiedNumber(retryAfterHeader)) {
+                this.retryAfter = parseInt(retryAfterHeader, 10);
+            }
+            else if (isStringifiedDate(retryAfterHeader)) {
+                // determine seconds till retry-after date
+                const timeTillRetrySeconds = Math.round(((new Date(retryAfterHeader) - Date.now()) / 1000));
+                // in case seconds till retry is negative, default to 1 second
+                this.retryAfter = timeTillRetrySeconds > 0? timeTillRetrySeconds : 1;
```
Comment (pheenomenon) [inline lib/error.js:52]: 1s for default will only happen for negative retry-after correct? else 1s might be too short.

File: lib/client-retry.js [hunk 24]
```diff
@@ -0,0 +1,151 @@
...
+            try {
+                response = await super.process(source, renditions, userData);
+                console.log('Successful call to /process', retryCount);
+                non429Response = true;
+            } catch (e) {
+                // string comparision instead of `instanceof` since `instanceof` causes issues
+                if (e.code === TOO_MANY_REQUEST_ERROR_CODE && e.name === TOO_MANY_REQUEST_ERROR) {
+                    non429Response = false;
+                    console.log(colors.red(`encountered a 429: ${e.message}.`));
+                    const waitTime = this.retryWaitTime(e.retryAfter);
```
Comment (pheenomenon) [inline lib/client-retry.js:128]: can `e.retryAfter` be undefined? good idea to add a test.

**Review by pheenomenon — APPROVED** [review]

File: lib/error.js (hunk 23 above)
Comment (jdelbick) [reply lib/error.js:52]: Correct. I could see this happening in a case where `retry-after` returned a date.
if the date is only 1s from the response, by the time we get to here, the date could be in the past, hence why I added this edge case

File: lib/client-retry.js (hunk 24 above)
Comment (jdelbick) [reply lib/client-retry.js:128]: I have a test at the micro level (asset-compute.js): https://github.com/adobe/asset-compute-client/blob/NUI-1650/test/assetcompute.test.js#L184-L208
I can add one at the client-retry.js level too

File: lib/client-retry.js (hunk 12 above)
Comment (jdelbick) [reply lib/client-retry.js]: I went with this option (abstract to function):
```
while (shouldRetry(non429Response, retryCount)) {
``

Comment (pheenomenon) [reply lib/client-retry.js]: I was thinking about `I am also trying to have a shared retry function` . excuse me if I confused with extracting out `while (!non429Response && retryCount <= this.max429RetryCount) {`

Was not sure if `super.register` can be passed in as a function object. There may be some Prototype trick here. Or else call a method passing in the name of the endpoint and use it to do super.(that endpoint).

```function callEndpointWithRetry(endpoint) {
  while loop
    if(endpoint === 'register'){
      super.register()
    }
  other code;
}
```

File: test/client-retry.test.js [hunk 25]
```diff
@@ -0,0 +1,858 @@
...
+    consumerId: 105979,
+    metascopes: ['event_receiver_api', 'ent_adobeio_sdk', 'asset_compute_meta'],
+    technicalAccount: {
+        id: 'id',
+        org: 'org',
+        clientId: 'clientId',
+        clientSecret: 'clientSecret',
+        privateKey: 'privateKey'
+    },
+    imsEndpoint: 'https://ims-na1-stage.adobelogin.com'
```
Comment (tmathern) [inline test/client-retry.test.js:34]: Could you use an obviously mocked URL for this test?

File: test/client-retry.test.js [hunk 26]
```diff
@@ -0,0 +1,858 @@
...
+
+const mockRequire = require('mock-require');
+const assert = require('assert');
+const nock = require('nock');
+const { EventEmitter } = require('events');
+
+const DEFAULT_INTEGRATION = {
+    applicationId: 72515,
+    consumerId: 105979,
+    metascopes: ['event_receiver_api', 'ent_adobeio_sdk', 'asset_compute_meta'],
```
Comment (tmathern) [inline test/client-retry.test.js:26]: Could you used different mocked metascopes for this test?

**Review by tmathern — CHANGES_REQUESTED** [review]

**Review by tmathern — APPROVED** [review]

File: lib/client-retry.js [hunk 27]
```diff
@@ -0,0 +1,170 @@
...
+     * 
+     * It will continue to retry if it has not received anything other than a 429 response 
+     * or retry count is up
+     * @param {boolean} non429Response false if received a 429 response
+     * @param {number} retryCount 
+     * @returns 
+     */
+    shouldRetry(non429Response, retryCount) {
+        return !non429Response && retryCount <= this.max429RetryCount;
+    }
```
[the remark this answers is not in the document]
Comment (bobvanmanen) [reply lib/client-retry.js:59]: Double negation is easy to misunderstand: `!non429Response`

File: lib/client-retry.js [hunk 28]
```diff
@@ -0,0 +1,170 @@
...
+        }
+        if (!non429Response) {
+            throw new Error(`Running into 429s after ${retryCount} attempts. Will not continue retrying. Try again in a few minutes.`);
+
+        }
+        // pass through error
+        if (error) {
+            throw error;
+        }
+        return response;
```
Comment (bobvanmanen) [inline lib/client-retry.js:92]: I think this may be easier to understand if you track `lastError`, set when caught and create a function that could tell you if an error is a `Too Many Request` error. Then replace the end with something like:

```
if (!lastError) {
return response;
}else if (tooManyRequestError(lastError)) {
throw new Error(`Running in to 429s...`);
} else {
throw lastError;
}
```

File: lib/client-retry.js [hunk 29]
```diff
@@ -0,0 +1,170 @@
...
+    }
```
Comment (bobvanmanen) [inline lib/client-retry.js:165]: Maybe it would be easier to create a retry function to contain the retry behavior so the duplication can be avoided. The function could be something like:

```
async function retryOnTooManyRequests(func) {
   //... retry behavior
   const response = await func();
   //... retry behavior
}
```

I did this here for http upload/download: https://github.com/adobe/node-httptransfer/blob/master/lib/retry.js#L169

The options passing is a bit messy, but adding retry is pretty straight forward: https://github.com/adobe/node-httptransfer/blob/master/lib/file.js#L103-L114

File: lib/error.js [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+
+"use strict";
+
+function isStringifiedDate(date) {
+    return (new Date(date) !== "Invalid Date") && !isNaN(new Date(date));
+}
+
+function isStringifiedNumber(num) {
+    const parsedRetryAfter = parseInt(num, 10);
+    return typeof(parsedRetryAfter) === "number" && !isNaN(parsedRetryAfter);
```
Comment (bobvanmanen) [inline lib/error.js:21]: I'm starting to like Number.isFinite for this check: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isFinite

It makes sure the value is a number, and the number is neither positive infinity, negative infinite or NaN.

File: lib/error.js [hunk 31]
```diff
@@ -0,0 +1,61 @@
...
+        super(message);
+
+        Error.captureStackTrace(this, TooManyRequestsError);
+        this.name = "TooManyRequestsError";
+        this.code = 429;
+
+        // `retry-after` header can be a stringified number in seconds or a stringified date
+        if (typeof(retryAfterHeader) === 'string') {
+            if (isStringifiedNumber(retryAfterHeader)) {
+                this.retryAfter = parseInt(retryAfterHeader, 10);
```
Comment (bobvanmanen) [inline lib/error.js:46]: This is parsing the number twice. Maybe instead have a parseNumber function that returns the result to avoid it?

Comment (bobvanmanen) [reply lib/error.js:46]: Not really concerned about performance, but rather code maintenance and the parseInt in `isStringifiedNumber` to get out of sync with this parseInt

File: lib/error.js (hunk 23 above)
Comment (bobvanmanen) [reply lib/error.js:52]: Same comment here, this is parsing the date twice. On second though maybe just have a `parseRetryAfter` which handles numbers and dates. You could write unit tests for that easily then as well.

**Review by bobvanmanen — CHANGES_REQUESTED** [review]

File: lib/client-retry.js [hunk 32]
```diff
@@ -0,0 +1,170 @@
...
+const TOO_MANY_REQUEST_ERROR = 'TooManyRequestsError';
+const TOO_MANY_REQUEST_ERROR_CODE = 429;
+/**
+ * Small wrapper around AssetComputeClient adding retry logic on 429s based on `retry-after` header
+ */
+class AssetComputeClientWithRetry extends AssetComputeClient {
+    constructor(integration, options) {
+        super(integration, options);
+        this.max429RetryCount = (options && options.max429RetryCount) || MAX_RETRY_ON_HTTP_429;
+    }
```
[the remark this answers is not in the document]
Comment (jdelbick) [reply lib/client-retry.js:30]: I forgot at this point why I made it a separate class. let me try to make it a configurable option

[the remark this answers is not in the document]
Comment (jdelbick) [reply lib/client-retry.js:30]: So the simplest option would be to pass in custom retry options to the `node-fetch-retry` request inside `assetcompute.js`, (https://github.com/adobe/asset-compute-client/blob/master/lib/assetcompute.js#L95)
something like this:
```
retryOnHttpResponse: function (response) {
        if (response.status === 429 || response.status >= 500) {
              return true;
        }
    }
```
However, 429s are a special case where the time to wait is in a header `retry-after`.
And `node-fetch-retry` does not have a way to dynamically change the wait time based on client input. We could add this..? thoughts?

File: lib/client-retry.js (hunk 29 above)
Comment (jdelbick) [reply lib/client-retry.js]: I took inspiration from you retry function in node-httptransfer and added something similar here: https://github.com/adobe/asset-compute-client/blob/NUI-1650/lib/retry.js

Now instead of a separate class, the 429 retry logic is integrated into `assetcompute.js`: https://github.com/adobe/asset-compute-client/blob/NUI-1650/lib/assetcompute.js#L93-L94

However, `assetcompute.js` already uses `node-fetch-retry` for retry logic. So, I kept the `retry` function only for custom 429 retry logic. and we will use `node-fetch-retry` for the rest of the retry logic: https://github.com/adobe/asset-compute-client/blob/NUI-1650/lib/assetcompute.js#L15

File: README.md [hunk 33]
```diff
@@ -153,10 +160,27 @@ await assetCompute.register();
...
+Retry 10 times:
+```js
+const assetCompute = new AssetComputeClient(integration, {
+    max429RetryCount: 10
+});
+```
+Disable retry:
+```js
+const assetCompute = new AssetComputeClient(integration, {
+    disable429Retry: 10
```
Comment (bobvanmanen) [inline README.md:175]: Should this be `true`?

**Review by bobvanmanen — APPROVED** [review]
LGTM, only have a doc comment and a few optimization comments that are not strictly required.

File: README.md (hunk 33 above)
Comment (jdelbick) [reply README.md]: yep, nice catch. I also made the optimization changes you recommended for the error class
